### PR TITLE
SONARJAVA-1563 RSPEC-2637 "@NonNull" values should not be set to null

### PR DESCRIPTION
**Pull request by DidierBesset**

### CODE REVIEW

File: java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java [hunk 1]
```diff
@@ -239,13 +240,21 @@ public void visitReturnStatement(ReturnStatementTree syntaxNode) {
...
       final ExpressionTree variable = syntaxNode.variable();
-      if (variable.is(Tree.Kind.ARRAY_ACCESS_EXPRESSION)) {
+      if (isNonLocalStorage(variable)) {
         List<SymbolicValue> stackedValues = programState.peekValues(2);
         SymbolicValue value = stackedValues.get(1);
         closeResource(value);
       }
     }
 
+    private boolean isNonLocalStorage(ExpressionTree variable) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/UnclosedResourcesCheck.java]: MAJOR Make "isNonLocalStorage" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+  private void checkVariable(CheckerContext context, MethodTree tree, VariableTree member) {
+    String nonNullAnnotation = nonNullAnnotation((JavaSymbol) member.symbol());
+    if (nonNullAnnotation != null && isUndefinedOrNull(context, member.symbol())) {
+      context.reportIssue(tree, this,
+        MessageFormat.format("\"{0}\" is marked \"{1}\" but is not initialized in this constructor.", member.symbol().name(), nonNullAnnotation));
+    }
+  }
+
+  private boolean isUndefinedOrNull(CheckerContext context, Symbol symbol) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: MAJOR Make "isUndefinedOrNull" a "static" method. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2325)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues:
* MAJOR 2 major

Watch the comments in this conversation to review them.

File: java-checks/src/main/java/org/sonar/java/checks/CheckList.java.orig [hunk 3]
```diff
@@ -0,0 +1,478 @@
+/*
```
Comment (m-g-sonar) [inline java-checks/src/main/java/org/sonar/java/checks/CheckList.java.orig]: Remove this file!

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+    public void visitReturnStatement(ReturnStatementTree tree) {
+      Tree parent = tree.parent();
+      while (!parent.is(Tree.Kind.METHOD)) {
+        parent = parent.parent();
+        if (parent == null) {
+          // This occurs when the return statement is within a constructor
+          return;
+        }
+      }
+      MethodTree methodTree = (MethodTree) parent;
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: MAJOR Rename "methodTree" which hides the field declared at line 60. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AHiddenFieldCheck)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/test/files/se/NonNullSetToNullCheck.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+  public MainClass(@Nonnull String color, String other) {
+    primary = color;
+    secondary = other;
+  }
+  
+  public MainClass() { // Noncompliant  {{"primary" is marked "javax.annotation.Nonnull" but is not initialized in this constructor.}}
+    return; // Just for coverage
+  }
+  
+  public @Nonnull String colorMix() {
```
Comment (m-g-sonar) [inline java-frontend/src/test/files/se/NonNullSetToNullCheck.java]: Usually we place the annotation related to the return type of the method before the modifiers. See for instance: https://github.com/SonarSource/sonar-java/blob/master/java-frontend/src/main/java/org/sonar/java/se/ProgramState.java#L327

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+      ClassTree classTree = (ClassTree) methodTree.parent();
+      for (Tree member : classTree.members()) {
+        if (member.is(Tree.Kind.VARIABLE)) {
+          checkVariable(context, methodTree, (VariableTree) member);
+        }
+      }
+    }
+  }
+
+  private void checkVariable(CheckerContext context, MethodTree tree, VariableTree member) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: As you only use the symbol related to the member, provide the symbol as argument, not the variableTree.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 7]
```diff
@@ -0,0 +1,237 @@
...
+    String nonNullAnnotation = nonNullAnnotation((JavaSymbol) member.symbol());
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: no need to cast to JavaSymbol.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+    }
+  }
+
+  private static boolean isUndefinedOrNull(CheckerContext context, Symbol symbol) {
+    ProgramState programState = context.getState();
+    SymbolicValue value = programState.getValue(symbol);
+    return value == null;
+  }
+
+  protected String nonNullAnnotation(JavaSymbol javaSymbol) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: Simply use a Symbol, `metadata().isAnnotatedWith()` is part of the API.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 9]
```diff
@@ -0,0 +1,237 @@
...
+  @Override
+  public ProgramState checkPreStatement(CheckerContext context, Tree syntaxNode) {
+    final AbstractStatementVisitor visitor = new PreStatementVisitor(context);
+    syntaxNode.accept(visitor);
+    return visitor.programState;
+  }
+
+  @Override
+  public ProgramState checkPostStatement(CheckerContext context, Tree syntaxNode) {
+    final PostStatementVisitor visitor = new PostStatementVisitor(context);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: why not use an `AbstractStatementVisitor` ? (and remove `final`...)

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+  protected String nonNullAnnotation(JavaSymbol javaSymbol) {
+    for (String annotation : ANNOTATIONS) {
+      if (javaSymbol.metadata().isAnnotatedWith(annotation)) {
+        return annotation;
+      }
+    }
+    return null;
+  }
+
+  private class AbstractStatementVisitor extends CheckerTreeNodeVisitor {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: If the class is called `AbstractStatementVisitor`, maybe it should be `abstract` too.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (tree.variable().is(Tree.Kind.IDENTIFIER)) {
+        IdentifierTree variable = (IdentifierTree) tree.variable();
+        JavaSymbol symbol = (JavaSymbol) variable.symbol();
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: no need to use `JavaSymbol`.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+        argumentValues.remove(0);
+        Collections.reverse(argumentValues);
+        JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) symbol;
+        checkNullArguments(syntaxTree, methodSymbol.getParameters(),
+          argumentValues, "Parameter {0} to this call is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    protected void checkNullArguments(Tree syntaxTree, Scope parameters, List<SymbolicValue> argumentValues, String message) {
+      if (parameters != null) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: Not covered by tests

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+        List<JavaSymbol> scopeSymbols = parameters.scopeSymbols();
+        int parametersToTest = argumentValues.size();
+        if (scopeSymbols.size() < parametersToTest) {
+          // The last parameter is a variable length argument: the non-null condition does not apply to its values
+          parametersToTest = scopeSymbols.size() - 1;
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: Not covered by tests.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+        JavaSymbol.MethodJavaSymbol methodSymbol = (JavaSymbol.MethodJavaSymbol) symbol;
+        checkNullArguments(syntaxTree, methodSymbol.getParameters(),
+          argumentValues, "Parameter {0} to this constructor is marked \"{1}\" but null is passed.");
+      }
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxTree) {
+      Symbol symbol = syntaxTree.symbol();
+      if (symbol.isMethodSymbol()) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: Case of unknown method not covered.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+
+  private class PreStatementVisitor extends AbstractStatementVisitor {
+
+    protected PreStatementVisitor(CheckerContext context) {
+      super(context);
+    }
+
+    @Override
+    public void visitAssignmentExpression(AssignmentExpressionTree tree) {
+      if (tree.variable().is(Tree.Kind.IDENTIFIER)) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: At least one other case than assignment to a variable should be covered by unit tests.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+    }
+
+    @Override
+    public void visitMethodInvocation(MethodInvocationTree syntaxTree) {
+      Symbol symbol = syntaxTree.symbol();
+      if (symbol.isMethodSymbol()) {
+        Arguments arguments = syntaxTree.arguments();
+        List<SymbolicValue> argumentValues = new ArrayList<>(programState.peekValues(arguments.size() + 1));
+        argumentValues.remove(0);
+        Collections.reverse(argumentValues);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: I would add a comment to at least distinguish the cases and explain why arguments should be reversed in method invocation when it's not the case for constructors.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+          return;
+        }
+      }
+      MethodTree methodTree = (MethodTree) parent;
+      String nonNullAnnotation = nonNullAnnotation((JavaSymbol) methodTree.symbol());
+      if (nonNullAnnotation != null) {
+        SymbolicValue returnedValue = programState.peekValue();
+        Constraint constraint = programState.getConstraint(returnedValue);
+        if (constraint == null) {
+          reportIssue(tree, "This method''s return value is marked \"{0}\" but null may be returned.", nonNullAnnotation);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: I'm not sure about reporting that issue. What about that case; The field is final and set in the constructor:

```
class MyOtherClass {
  private final Object myObject;
  
  public MyOtherClass() {
    myObject = new Object();
  }
  
  @Nonnull
  public Object getMyObject() {
    return myObject; // would raise a FP
  }
}
```

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+        parent = parent.parent();
+        if (parent == null) {
+          // This occurs when the return statement is within a constructor
+          return;
+        }
+      }
+      MethodTree mTree = (MethodTree) parent;
+      String nonNullAnnotation = nonNullAnnotation(mTree.symbol());
+      if (nonNullAnnotation != null) {
+        if (isLocalExpression(tree.expression())) {
```
Comment (SonarTech) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: MAJOR Merge this if statement with the enclosing one. [rule](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS1066)

**Comment by SonarTech** [comment]
SonarQube analysis reported 1 issue:
* MAJOR 1 major

Watch the comments in this conversation to review them.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+          checkReturnedValue(tree, nonNullAnnotation);
+        }
+      }
+    }
+
+    private boolean isLocalExpression(ExpressionTree expression) {
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: Looks very close from the method `isNonLocalStorage` that you implemented in 'UnclosedResourcesCheck', except that this time you are checking if it's a local variable. The two methods should be renamed to match this behavior.

File: java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+        return symbol.isMethodSymbol();
+      }
+      return true;
+    }
+
+    private void checkReturnedValue(ReturnStatementTree tree, String nonNullAnnotation) {
+      SymbolicValue returnedValue = programState.peekValue();
+      Constraint constraint = programState.getConstraint(returnedValue);
+      if (constraint == null) {
+        reportIssue(tree, "This method''s return value is marked \"{0}\" but null may be returned.", nonNullAnnotation);
```
Comment (m-g-sonar) [inline java-frontend/src/main/java/org/sonar/java/se/checks/NonNullSetToNullCheck.java]: We should not raise issues in this case, as we are not certain. As we are not (yet) doing cross-procedural analysis, doing the assumption that we may end up with a null value with necessarily raise à lot of FPs.

**Pull request merged**